contentcuration/contentcuration/db/models/manager.py

```diff
@@ -74,54 +122,55 @@ def _move_child_to_new_tree(self, node, target, position):
             Q(prerequisite_id=node.id) | Q(target_node_id=node.id)
         ).delete()
 
-    @contextlib.contextmanager
-    def _update_changes(self, node, save):
-        original_parent_id = node.parent_id
-        yield
-        ids = [original_parent_id, node.parent_id] + [node.id] if save else []
-        # Always write to the database for the parent change updates, as we have
-        # no persistent object references for the original and new parent to modify
-        if ids:
-            self.filter(id__in=ids).update(changed=True)
-        node.changed = True
-
-    def _move_node(
-        self, node, target, position="last-child", save=True, refresh_target=True
-    ):
-        # If we are delaying updates, then _move_node defers to insert_node
-        # we are already wrapping for parent changes below, so no need to
-        # add our parent changes context manager.
-        if self.tree_model._mptt_is_tracking:
-            return super(CustomContentNodeTreeManager, self)._move_node(
-                node, target, position, save, refresh_target
-            )
-        with self._update_changes(node, save):
-            return super(CustomContentNodeTreeManager, self)._move_node(
-                node, target, position, save, refresh_target
-            )
-
-    def insert_node(
-        self,
-        node,
-        target,
-        position="last-child",
-        save=False,
-        allow_existing_pk=False,
-        refresh_target=True,
-    ):
-        with self._update_changes(node, save):
-            if save:
-                with self.lock_mptt(target.tree_id):
-                    return super(CustomContentNodeTreeManager, self).insert_node(
-                        node, target, position, save, allow_existing_pk, refresh_target
-                    )
-            return super(CustomContentNodeTreeManager, self).insert_node(
-                node, target, position, save, allow_existing_pk, refresh_target
+    def _mptt_refresh(self, *nodes):
```

Comment:
Could you explain why this code is needed? It appears to be getting the node's mptt attrs from the db, and then setting them on the object. Is this because they are missing from the node object being passed in, or are the values potentially incorrect?

Reply:
It's possible that the node values are incorrect - wherever the node data was loaded, another process may now have caused an update to the tree that these nodes are members of, and led these nodes to have stale mptt values.

This function is called once inside a lock of the mptt fields, so when this is called, the trees that these nodes come from are now locked for updates, so the values will not go stale during the course of the lock.

Reply:
Gotcha, thanks!